aten/src/ATen/native/xnnpack/Pooling.cpp

```diff
@@ -0,0 +1,321 @@
+
+  return Context{
+    Operator(max_pool_op),
+    {
+      output_channels,
+      {
+        kernel[Layout::Parameter::height],
+        padding[Layout::Parameter::height],
+        stride[Layout::Parameter::height],
+        dilation[Layout::Parameter::height],
+        ceil_mode,
+      },
+      {
+        kernel[Layout::Parameter::width],
+        padding[Layout::Parameter::width],
+        stride[Layout::Parameter::width],
+        dilation[Layout::Parameter::width],
+        ceil_mode,
+      },
+    }
+  };
+}
+
+bool shape(const int64_t input, const Context::Output::Spatial& spatial) {
+  return pooling_output_shape(
+      input,
+      spatial.kernel,
+      spatial.padding,
+      spatial.stride,
+      spatial.dilation,
+      spatial.ceil_mode);
+};
+
+bool usable(
+    const Tensor& input,
+    const Context::Output::Spatial& height,
+    const Context::Output::Spatial& width) {
+  using namespace internal;
+
+         // Input
+  return (4 == input.dim()) &&
+         (c10::DeviceType::CPU == input.device().type()) &&
+         (kFloat == input.scalar_type()) &&
+         // Output
+         (shape(input.size(Layout::Activation4D::height), height) > 0) &&
+         (shape(input.size(Layout::Activation4D::width), width) > 0) &&
+         true;
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  TORCH_CHECK(
+      usable(input, context.output.height, context.output.width),
+      "mobile::cpu::max_pool not usable!");
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+  Tensor output_nhwc = empty_with_tail_padding(
+      {
+        input_nhwc.size(Layout::Activation4D::batch),
+        context.output.channels,
+        shape(input_nhwc.size(Layout::Activation4D::height), context.output.height),
+        shape(input_nhwc.size(Layout::Activation4D::width), context.output.width),
+      },
+      input_nhwc.options().dtype(),
+      MemoryFormat::ChannelsLast);
+
+  const xnn_status setup_status = xnn_setup_max_pooling2d_nhwc_f32(
+      context.max_pool_op.get(),                      // operator
+      input_nhwc.size(Layout::Activation4D::batch),   // batch_size
+      input_nhwc.size(Layout::Activation4D::height),  // input_height
+      input_nhwc.size(Layout::Activation4D::width),   // input_width
+      input_nhwc.data_ptr<float>(),                   // input
+      output_nhwc.data_ptr<float>(),                  // output
+      nullptr);                                       // threadpool
+
+  TORCH_CHECK(
+      xnn_status_success == setup_status,
+      "xnn_setup_max_pooling2d_nhwc_f32 failed!");
+
+  const xnn_status run_status = xnn_run_operator(
+      context.max_pool_op.get(),  // operator
+      nullptr);                   // threadpool
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == run_status,
+      "xnn_run_operator failed!");
+
+  return output_nhwc.contiguous(input.suggest_memory_format());
```

Comment:
Why not just return output_nhwc?

Reply:
I think PyTorch's convention is to return tensors in the same layout they came in so we'll have to switch back to NCHW if that's the layout the input tensor was in.  Dima / Natalia can confirm.

Reply:
If that's the convention, then it makes sense to stick with it.

We should make sure that if we're returning NHWC, we're doing so with a guarding allocator to get maximum performance from a sequence of XNNPACK ops.

Reply:
contiguous is a no-op if memory is already in the requested layout.  In other words, if the input tensor is in NHWC, contiguous will short circuit..